Tree Borrows without protector (Path Provenance) (#220)
minirust/minirust: spec/mem/tree_borrows/memory.md

        // Create the root node for the tree.
        // Initially, we set the permission as `Active`
        let root = Node {
            children: List::new(),
            location_states: Self::init_location_states(Permission::Active, size),
        };

        // Path to the root node
        let path = List::new();

        let allocation = Allocation {
            addr,
            align,
            kind,
            live: true,
            data: list![AbstractByte::Uninit; size.bytes()],
            extra: TreeBorrowsAllocationExtra { root },
        };

        self.allocations.push(allocation);

        ret(Pointer { addr, provenance: Some((path, alloc_id)) })
    }
}

impl<T: Target> Memory for TreeBorrowsMemory<T> {
    fn deallocate(&mut self, ptr: Pointer<Self::Provenance>, kind: AllocationKind, size: Size, align: Align) -> Result {
        let Some((path, alloc_id)) = ptr.provenance else {
            throw_ub!("deallocating invalid pointer")
        };
        // This lookup will definitely work, since AllocId cannot be faked.
        let mut allocation = self.allocations[alloc_id.0];

        allocation.deallocation_check(ptr.addr, kind, size, align)?;

>         // Check that ptr has the permission to write the entire allocation

Reviewer: ```suggestion
        // Check that ptr has the permission to write the entire allocation.
```

        allocation.extra.root.access(path, AccessKind::Write, Size::ZERO, size, true)?;

        // Mark it as dead. That's it.
        self.allocations.mutate_at(alloc_id.0, |allocation| {
            allocation.live = false;
        });

        ret(())
    }
}
```

### Load Operation

> Corresponding to the `AccessKind::Read`.

Reviewer: I don't understand the point of this comment -- AccessKind::Read is an internal Tree Borrows thing, so not sure why it would be called out here?

I think you can have alloc, dealloc, load, store in one block and just say above that this is defining the core memory operations.

```rust
impl<T: Target> Memory for TreeBorrowsMemory<T> {
    fn load(&mut self, ptr: Pointer<Self::Provenance>, len: Size, align: Align) -> Result<List<AbstractByte<Self::Provenance>>> {
       let Some((path, alloc_id, offset)) = self.check_ptr(ptr, len)? else {
            return ret(list![]);
        };

>         // Recursively update the tree and check the existence of UBs 
>         let mut allocation = self.allocations[alloc_id.0];
> 
>         // Load the data
>         let data = allocation.load(ptr.addr, offset, len, align)?;
> 
>         allocation.extra.root.access(path, AccessKind::Read, offset, len, true)?;

Reviewer: ```suggestion
        let mut allocation = self.allocations[alloc_id.0];

        // Check for aliasing violations.
        allocation.extra.root.access(path, AccessKind::Read, offset, len, true)?;
        // Load the data.
        let data = allocation.load(ptr.addr, offset, len, align)?;
```
Can this be done with `mutate_at`?
